src/os/bluestore/BlueFS.cc

```diff
@@ -3133,16 +3126,106 @@ bool BlueFS::wal_is_rotational()
   return bdev[BDEV_SLOW]->is_rotational();
 }
 
+#undef dout_prefix
+#define dout_prefix *_dout << "bluefs rebalance "
+
 void BlueFS::consider_move()
 {
-  //if currently moving something, ignore
+  //this function does make sense only if both BDEV_DB and BDEV_SLOW are used
+  /*
+   * There are 3 modes of operation for moving bluefs files, determined by % of db used.
+   * %0
+   * A | move db_on_slow->db_on_db
+   * A | move slow_on_slow->slow_on_db
+   * %all_on_db
+   * B | move db_on_slow->db_on_db
+   * %no_slow_on_db
+   * C | move slow_on_db->slow_on_slow
+   * %100
+   *
+   * To prevent circular operation distance ( %all_on_db - %no_slow_on_db ) must be larger then largest file.
+   */
+  utime_t now = ceph_clock_now();
+  if (double(move_next_check - now) > 0) {
+    return;
+  }
+
+  std::unique_lock l(lock);
+  if (move_context)
+    return;
+
+  double all_on_db = cct->_conf.get_val<double>("bluestore_bluefs_placement_all_on_db");
+  double no_slow_on_db = cct->_conf.get_val<double>("bluestore_bluefs_placement_no_slow_on_db");
+
+  double frac = 1 - (double) alloc[BDEV_DB]->get_free() / (double) block_all[BDEV_DB].size();
+  size_t size_on_slow = block_all[BDEV_SLOW].size() - alloc[BDEV_SLOW]->get_free();
+  FileRef file;
+  if (frac < all_on_db) {
+    if (size_on_slow > 0) {
+      file = find_move_candidate(BDEV_SLOW, BDEV_DB);
+      if (file) {
+        file_move_start(file, BDEV_DB);
+      } else {
+        file = find_move_candidate(BDEV_SLOW, BDEV_SLOW);
+        if (file) {
+          file_move_start(file, BDEV_DB);
+        }
+      }
+    }
+  } else if (frac < no_slow_on_db) {
+    if (size_on_slow > 0) {
+      file = find_move_candidate(BDEV_SLOW, BDEV_DB);
+      if (file) {
+        file_move_start(file, BDEV_DB);
+      }
+    }
+  } else {
+    file = find_move_candidate(BDEV_DB, BDEV_SLOW);
+    if (file) {
+      file_move_start(file, BDEV_SLOW);
+    }
+  }
+}
+
+BlueFS::FileRef BlueFS::find_move_candidate(uint8_t is_on, uint8_t wants_on)
+{
+  mempool::bluefs::unordered_map<uint64_t,FileRef>::iterator it;
+  it = file_map.find(move_last_ino);
+  if (it != file_map.end())
+    it++;
+  if (it == file_map.end())
+    it = file_map.begin();
+  if (it == file_map.end()) {
+    dout(1) << __func__ << " no files" << dendl;
+    return FileRef{};
+  }
   
+  for (uint64_t cnt = 0; cnt < 1000; cnt++) {
+    bluefs_fnode_t& fnode = it->second->fnode;
+    if (fnode.prefer_bdev == wants_on &&
+        fnode.extents.size() != 0 &&
+        fnode.extents[0].bdev == is_on) {
```

Comment:
it's not enough to check extens[0] - file might have multiple extents at different devices simultaneously.